### PR TITLE
Research group can sign up with key

### PR DESCRIPTION
**Pull request by erikwesterberg**
PivotalTracker - https://www.pivotaltracker.com/story/show/167917098

- Refactor registrations_spec.rb so that it only tests for User with University role.

- Request spec for sign up user: research group role should have a key with many happy and sad paths.

- Refactor Registration Controller to only generate keys for user: university role.

- Refactor associations for the roles of User in the model, research group and university with "classname" user.

- Generate migration file to make associaton between university-user and research group-user

- Make instance associations between User: university role and User: research group role when registration_key is used

- Merge latest upstream and resolve conflicts

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -5,7 +5,7 @@ class ApplicationController < ActionController::API
   protected
 
   def configure_permitted_parameters
-    devise_parameter_sanitizer.permit(:sign_up, keys: [:name, :role])
+    devise_parameter_sanitizer.permit(:sign_up, keys: [:name, :role, :university_id])
```
Comment (tochman) [inline app/controllers/application_controller.rb]: Why include `university_id` in this sanitation?

File: app/controllers/registrations_controller.rb [hunk 2]
```diff
@@ -1,11 +1,93 @@
 class RegistrationsController < ::DeviseTokenAuth::RegistrationsController
-  def render_create_success
-    5.times { RegistrationKey.create(user: @resource) }
-    @resource.reload    
+  
+  def create
+
+    build_resource
+    if params[:role] === 'research_group'
```
Comment (tochman) [inline app/controllers/registrations_controller.rb]: We don't use triple = in ruby

File: app/controllers/registrations_controller.rb [hunk 3]
```diff
@@ -1,11 +1,93 @@
...
+
+      unless reg_key.nil?
+        @resource.university = reg_key.user 
+      else
+        render_json_error_response('Invalid registration key') and return
+      end
+    end 
+  
+
+    # every line of code for the rest of the create action is devise code
```
Comment (tochman) [inline app/controllers/registrations_controller.rb:21]: Please create a refactoring ticket for this. We want to be able to use `super` here, but I understand that we loose the `.university` in the process. So we can leave it like this for now, and revisit this if there's time.

File: app/controllers/registrations_controller.rb [hunk 4]
```diff
@@ -1,11 +1,93 @@
...
-      data: resource_data.merge(registration_keys: @resource.registration_keys)
+      registration_keys: @resource.registration_keys,
+      data: resource_data
     }
   end
+
+  def render_json_error_response(message)
+    render json: {
+      status: 'error',
+      data:   resource_data,
```
Comment (tochman) [inline app/controllers/registrations_controller.rb:87]: Why do you include `resource_data` in an error response?

File: spec/requests/api/v0/research_group_sign_up_with_key_spec.rb [hunk 5]
```diff
@@ -0,0 +1,79 @@
+RSpec.describe 'Registration', type: :request do
+  let(:header) { { HTTP_ACCEPT: 'application/json' } }
+  let(:registration_key) { 'o7A8pJcuvzhv7fih9Paak3nt' }
```
Comment (tochman) [inline spec/requests/api/v0/research_group_sign_up_with_key_spec.rb]: Why do you create this? I can't see that you use this for anything.

File: spec/requests/api/v0/research_group_sign_up_with_key_spec.rb [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  let(:university) { FactoryBot.create(:user, role: 'university') }
```
Comment (tochman) [inline spec/requests/api/v0/research_group_sign_up_with_key_spec.rb]: Please stick to the shorter syntax for creating objects with FactoryBot

File: spec/requests/api/v0/research_group_sign_up_with_key_spec.rb [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+  describe 'of User with Research Group role with valid Registration Key' do
+    before 'post new User info' do
+      post '/api/v0/auth', params: { email: 'example@craftacademy.se',
+                                    name: 'Research Group Alpha',
+                                    role: 'research_group',
+                                    password: 'password',
+                                    password_confirmation: 'password',
+                                    registration_key: reg_key.combination },
+                                    headers: headers
```
Comment (tochman) [inline spec/requests/api/v0/research_group_sign_up_with_key_spec.rb:14]: The variable you create on line 2 is called `header` yet here you are referencing `headers`???

File: spec/requests/api/v0/research_group_sign_up_with_key_spec.rb [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+    it 'verifies that User with Research Group role is created' do
+      expect(response_json["data"]["name"]).to eq 'Research Group Alpha'
+    end
+
+    it 'verifies that created University has a Registration key' do
+      expect(response_json["data"]["university_id"]).to eq reg_key.user_id
+    end
+    
+    it 'verifes that Research group are associated with the University thats created' do
```
Comment (tochman) [inline spec/requests/api/v0/research_group_sign_up_with_key_spec.rb]: This is a unit spec and does not fit in this spec. Also, you are already testing this on line 27. I see the value of having this spec during development (as so-called " sanity check"), but it's not relevant here.

File: spec/requests/api/v0/research_group_sign_up_with_key_spec.rb [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+      post '/api/v0/auth', params: { email: 'maria@craftacademy.se',
+                                     name: 'Fat Jesus',
+                                     role: 'research_group',
+                                     password: 'password',
+                                     password_confirmation: 'password',
+                                     registration_key: nil },
+                                     headers: headers
+    end
+
+    it 'returns a 422 response, save successfully attempted but nothing saved' do
```
Comment (tochman) [inline spec/requests/api/v0/research_group_sign_up_with_key_spec.rb]: Not sure I understand this message. 422 is Unprocessable Entity

File: spec/requests/api/v0/research_group_sign_up_with_key_spec.rb [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+      expect(response.status).to eq 422
+    end
+
+    it 'returns error message' do
+      expect(response_json["errors"]).to eq 'Need a registration key'
```
Comment (tochman) [inline spec/requests/api/v0/research_group_sign_up_with_key_spec.rb]: This message feels a bit cryptic...

File: spec/requests/api/v0/university_registrations_spec.rb [hunk 11]
```diff
@@ -20,12 +20,24 @@
...
     end
+  end
 
-    it 'JSON body response contains a name ' do
-      expect(response_json['data']['name']).to eq 'Fat Bob'
+  describe 'with valid credentials for Research Group role' do
+    before 'post new Research Group info' do
+      post '/api/v0/auth', params: { email: 'example@craftacademy.se',
+                                    name: 'Fat Bob',
+                                    role: 'research_group',
```
Comment (tochman) [inline spec/requests/api/v0/university_registrations_spec.rb]: After  changes, this spec is about registering a `university`, yet here you are registering a `research_group`

File: spec/requests/api/v0/university_registrations_spec.rb [hunk 12]
```diff
@@ -20,12 +20,24 @@
...
+                                    password: 'password',
+                                    password_confirmation: 'password' },
+                                    headers: headers
     end
+
+    it 'returns zero registration keys' do
+      expect(response_json['data']['registration_keys'].count).to eq 0
```
Comment (tochman) [inline spec/requests/api/v0/university_registrations_spec.rb]: Is the purpose of the entire `describe` to make sure that we don't generate registration keys? In that case, you probably want to make sure that you actually use the new way of registering a `reserch_group` - meaning that you have to use a key to connect it to a university. Does that make sense? This spec actually exposes a flaw in your implementation allowing a `reserch_group` to register without a key.

**Review by tochman — CHANGES_REQUESTED** [review]

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (erikwesterberg) [reply app/controllers/application_controller.rb]: Removed

File: app/controllers/registrations_controller.rb (hunk 2 above)
Comment (erikwesterberg) [reply app/controllers/registrations_controller.rb]: bad habit, fixed

File: app/controllers/registrations_controller.rb (hunk 4 above)
Comment (erikwesterberg) [reply app/controllers/registrations_controller.rb:87]: We did not wanted to change the way devise token auth render error messages. 

here u got link - https://github.com/lynndylanhurley/devise_token_auth/blob/master/app/controllers/devise_token_auth/registrations_controller.rb

File: spec/requests/api/v0/research_group_sign_up_with_key_spec.rb (hunk 5 above)
Comment (erikwesterberg) [reply spec/requests/api/v0/research_group_sign_up_with_key_spec.rb]: Removed

File: app/controllers/registrations_controller.rb (hunk 3 above)
Comment (ghost) [reply app/controllers/registrations_controller.rb:21]: Done
https://www.pivotaltracker.com/story/show/168073036

File: spec/requests/api/v0/research_group_sign_up_with_key_spec.rb (hunk 6 above)
Comment (erikwesterberg) [reply spec/requests/api/v0/research_group_sign_up_with_key_spec.rb]: fixed!

File: spec/requests/api/v0/research_group_sign_up_with_key_spec.rb (hunk 7 above)
Comment (erikwesterberg) [reply spec/requests/api/v0/research_group_sign_up_with_key_spec.rb:14]: fixed typo!

File: spec/requests/api/v0/research_group_sign_up_with_key_spec.rb (hunk 8 above)
Comment (erikwesterberg) [reply spec/requests/api/v0/research_group_sign_up_with_key_spec.rb]: Fixed, test is now refactored!

File: spec/requests/api/v0/research_group_sign_up_with_key_spec.rb (hunk 9 above)
Comment (erikwesterberg) [reply spec/requests/api/v0/research_group_sign_up_with_key_spec.rb]: Fixed!

File: spec/requests/api/v0/research_group_sign_up_with_key_spec.rb (hunk 10 above)
Comment (erikwesterberg) [reply spec/requests/api/v0/research_group_sign_up_with_key_spec.rb]: Fixed!

File: spec/requests/api/v0/university_registrations_spec.rb (hunk 11 above)
Comment (erikwesterberg) [reply spec/requests/api/v0/university_registrations_spec.rb]: True , added test for registration keys in the research_group spec test.
removed this one

File: spec/requests/api/v0/university_registrations_spec.rb (hunk 12 above)
Comment (erikwesterberg) [reply spec/requests/api/v0/university_registrations_spec.rb]: Removed !

File: app/controllers/registrations_controller.rb (hunk 4 above)
Comment (tochman) [reply app/controllers/registrations_controller.rb:87]: Okay. I understand

**Review by julieahlquist — COMMENTED** [review]
Looks fine to me!

**Review by faraznaeem — APPROVED** [review]

**Pull request merged**